### PR TITLE
[CARBONDATA-3640] Insert from stage command support partition table

### PR DESCRIPTION
**Pull request by niuge01**
This feature is actually implemented by the CarbonDataLoadCommand.

 - [ ] Any interfaces changed?
NO
 
 - [ ] Any backward compatibility impacted?
 NO

 - [ ] Document update required?
NA

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
  YES
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1322/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1345/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1332/

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala [hunk 1]
```diff
@@ -189,4 +194,20 @@ class TestCarbonPartitionWriter extends QueryTest {
     output.asScala
   }
 
+  private def delDir(dir: File): Boolean = {
+    if (dir.isDirectory) {
+      val children = dir.list
+      if (children != null) {
+        var i = 0
+        while (i < children.length) {
```
Comment (jackylk) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala]: declare a val like `val len = children.length`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala [hunk 2]
```diff
@@ -269,11 +281,90 @@ case class CarbonInsertFromStageCommand(
...
+  private def startLoadingWithPartition(
+      spark: SparkSession,
+      table: CarbonTable,
+      loadModel: CarbonLoadModel,
+      stageInput: Seq[StageInput]
+    ): Unit = {
+    val partitionDataList = listPartitionFiles(stageInput)
+    val start = System.currentTimeMillis()
+    var index = 0
+    partitionDataList.map(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: formatted like:
```
partitoinDataList.map { partitionData =>
    ...
}

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala [hunk 3]
```diff
@@ -269,11 +281,90 @@ case class CarbonInsertFromStageCommand(
...
+          internalOptions = Map.empty,
+          partition = partition
+        )
+        loadCommand.run(spark)
+      }
+    )
     LOGGER.info(s"finish data loading, time taken ${System.currentTimeMillis() - start}ms")
   }
 
+  private def listPartitionFiles(stageInputs : Seq[StageInput]):
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: move `:` to next line

Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: describe the return tuple in comment

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala [hunk 4]
```diff
@@ -269,11 +281,90 @@ case class CarbonInsertFromStageCommand(
...
+    val partitionDataList = listPartitionFiles(stageInput)
+    val start = System.currentTimeMillis()
+    var index = 0
+    partitionDataList.map(
+      partitionData => {
+        index = index + 1
+        val partition = partitionData._1
+        val splits = partitionData._2
+        LOGGER.info(s"start to load ${splits.size} files into " +
+          s"${table.getDatabaseName}.${table.getTableName}")
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: suggest mention it is for which partition in the log

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala [hunk 5]
```diff
@@ -269,11 +281,90 @@ case class CarbonInsertFromStageCommand(
...
 
+  private def listPartitionFiles(stageInputs : Seq[StageInput]):
+  Seq[(Map[String, Option[String]], Seq[InputSplit])] = {
+    val partitionMap = new util.HashMap[Map[String, Option[String]], util.List[InputSplit]]()
+    stageInputs.foreach(
+      stageInput => {
+        val locations = stageInput.getLocations.asScala
+        locations.foreach(
+          location => {
+            val partition = location.getPartitions.asScala.map(t => (t._1, Option(t._2))).toMap
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: Why Option is required? Can't we remove it?

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala [hunk 6]
```diff
@@ -794,8 +792,8 @@ case class CarbonLoadDataCommand(
         sparkSession,
         operationContext)
       val logicalPlan = if (sortScope == SortScopeOptions.SortScope.GLOBAL_SORT) {
-        var numPartitions =
-          CarbonDataProcessorUtil.getGlobalSortPartitions(carbonLoadModel.getGlobalSortPartitions)
+        var numPartitions = 1
+        //  CarbonDataProcessorUtil.getGlobalSortPartitions(carbonLoadModel.getGlobalSortPartitions)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala]: remove it

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala [hunk 7]
```diff
@@ -898,12 +896,32 @@ case class CarbonLoadDataCommand(
...
-    val updatedRdd = convertData(
+    val convertedRdd = convertData(
       rdd,
       sparkSession,
       loadModel,
       isDataFrame,
       partitionValues)
+    val updatedRdd = if (isDataFrame) {
+      val columnCount = loadModel.getCsvHeaderColumns.length
+      convertedRdd.map{ row =>
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala]: add space before `{`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala [hunk 8]
```diff
@@ -375,5 +466,37 @@ case class CarbonInsertFromStageCommand(
...
+      columnProjection = new CarbonProjection(columns),
+      null,
+      carbonTable.getAbsoluteTableIdentifier,
+      carbonTable.getTableInfo.serialize,
+      carbonTable.getTableInfo,
+      new CarbonInputMetrics,
+      null,
+      null,
+      classOf[SparkRowReadSupportImpl],
+      splits.asJava)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: move `)` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala (hunk 2 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala (hunk 4 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala (hunk 5 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: The CarbonLoadDataCommand required Option, so we cann't remove it here.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala (hunk 6 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala (hunk 7 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala (hunk 8 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala]: done

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1337/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1327/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1350/

**Comment by niuge01** [comment]
please test this

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1331/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1341/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1354/

**Comment by jackylk** [comment]
retest this please

**Comment by niuge01** [comment]
retest this please

**Comment by niuge01** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1339/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1337/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1341/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1351/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1363/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**